API Endpoint - DagRuns Batch

takunnithan opened the pull request:
Partially Closes: #9112 

Get all DAG Runs from all DAGs.

`POST /dags/~/dagRuns/list`
This endpoint is a POST to allow filtering across a large number of DAG IDs, where as a GET it would run in to maximum HTTP request URL length limits.




---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

takunnithan committed:
wip: Code refactor & Updated get_dag_runs_batch()

takunnithan committed:
Added test case

ephraimbuddy commented on tests/api_connexion/endpoints/test_dag_run_endpoint.py:
```diff
@@ -346,6 +346,55 @@ def test_end_date_gte_lte(self, url, expected_dag_run_ids, session):
+            }
+        )
+
+
```
I would suggest you add more tests. Like testing the page limit and offset, the execution date etc. You may use parameterized for the tests

takunnithan replied:
Thanks! @ephraimbuddy .  I will add more test soon.

@ephraimbuddy @mik-laj Could you help me figure out why my test case is failing ? 

I am getting this error:
**TypeError: get_dag_runs_batch() missing 1 required positional argument: 'dag_ids'**

More details is here: https://github.com/apache/airflow/pull/9556/checks?check_run_id=816283026

I get the same error in local dev env. 

Thanks!

mik-laj replied:
Parameters for the view function are only passed if the parameters are from a URL  In this case, the parameters are passed as the body, so you need to handle them in the view.
https://connexion.readthedocs.io/en/latest/request.html#automatic-parameter-handling

takunnithan replied:
Thanks @mik-laj

ephraimbuddy commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -67,40 +67,72 @@ def get_dag_runs(session, dag_id, start_date_gte=None, start_date_lte=None,
+@provide_session
+def get_dag_runs_batch(session, dag_ids, start_date_gte=None, start_date_lte=None,
+                       execution_date_gte=None, execution_date_lte=None,
+                       end_date_gte=None, end_date_lte=None, offset=None, limit=None):
```
```suggestion
def get_dag_runs_batch(session):
```

ephraimbuddy commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -67,40 +67,72 @@ def get_dag_runs(session, dag_id, start_date_gte=None, start_date_lte=None,
+                       end_date_gte=None, end_date_lte=None, offset=None, limit=None):
     """
     Get list of DAG Runs
     """
```
```
body = request.get_json()
try:
     data = list_dag_runs_form_schema.load(body)
except ValidationError as err:
     raise BadRequest(detail=err.messages)
```

ephraimbuddy replied:
Your schema should be like this:

```
class ListDagRunsFormSchema(Schema):
    """ Schema for ListDagRunsForm """

    class Meta:
        """ Meta """
        dateformat = 'iso'

    page_offset = fields.Int(min=0)
    page_limit = fields.Int(min=1, max=100)
    dag_ids = fields.List(fields.Str())
    execution_date_gte = fields.DateTime()
    execution_date_lte = fields.DateTime()
    start_date_gte = fields.DateTime()
    start_date_lte = fields.DateTime()
    end_date_gte = fields.DateTime()
    end_date_lte = fields.DateTime()
```
then instantiate it as `list_dag_runs_form_schema = ListDagRunsFormSchema(strict=True)`

ephraimbuddy replied:
You have to add the schema at dag_runs_schema.py

takunnithan committed:
1. Code refactor based on PR comments
   a. Using Marshmallow schema to validate & deserialize request payload
2. Test fix

takunnithan committed:
Merge branch 'master' into issue-9112

takunnithan committed:
Added more test cases

ephraimbuddy commented on airflow/api_connexion/schemas/dag_run_schema.py:
```diff
@@ -72,5 +72,25 @@ class DAGRunCollectionSchema(Schema):
+    page_offset = fields.Int(required=False, missing=0, min=0)
+    page_limit = fields.Int(required=False, missing=100, min=1)
+    dag_ids = fields.List(fields.Str(), required=False, missing=None)
+    execution_date_gte = fields.DateTime(required=False, missing=None)
```
```suggestion
    execution_date_gte = fields.DateTime()
```
I suggest you remove the missing and required args.  To only use them when the value should be something other than False and None respectively. This is already the default from marshmallow. 
https://marshmallow.readthedocs.io/en/2.x-line/api_reference.html#module-marshmallow.fields

takunnithan replied:
Thanks @ephraimbuddy . I have removed `required=False` from the args. But without `missing=None`, the fields were missing in deserialized data.

OmairK commented on airflow/api_connexion/schemas/dag_run_schema.py:
```diff
@@ -72,5 +72,25 @@ class DAGRunCollectionSchema(Schema):
+        strict = True
+
+    page_offset = fields.Int(required=False, missing=0, min=0)
+    page_limit = fields.Int(required=False, missing=100, min=1)
```
```suggestion
    page_limit = fields.Int(required=False, missing=100, min=1, max=100)
```

It would be better to cap the limit to 100. @ephraimbuddy @takunnithan what do you think?

ephraimbuddy replied:
I think it's better left out. Some servers can handle more than 1000 page request limit. I'm working on configurable page request limit and we can apply it here. Then it'll not be a hard limit but a limit that the users can configure.
You can see the commit here https://github.com/apache/airflow/commit/a37ac22c179e5dccc95b05c862bffc542dad125e

turbaszek commented on tests/api_connexion/endpoints/test_dag_run_endpoint.py:
```diff
@@ -346,6 +346,243 @@ def test_end_date_gte_lte(self, url, expected_dag_run_ids, session):
+    def test_should_respond_200(self, session):
+        dag_runs = self._create_test_dag_run()
+        session.add_all(dag_runs)
+        session.commit()
```
```suggestion

    def test_should_respond_200(self):
        dag_runs = self._create_test_dag_run()
        with create_session() as session:
            session.add_all(dag_runs)
```
WDYT?

turbaszek replied:
Or maybe it would be better to extend `_create_test_dag_run` with commit option? For example `_create_test_dag_run(commit=True)` so we can avoid code duplication

takunnithan replied:
Thanks @turbaszek . This is addressed in the latest commit.

turbaszek commented on tests/api_connexion/endpoints/test_dag_run_endpoint.py:
```diff
@@ -346,6 +346,243 @@ def test_end_date_gte_lte(self, url, expected_dag_run_ids, session):
+        }
+        response = self.client.post("api/v1/dags/~/dagRuns/list", json=payload)
+        assert response.status_code == 200
+        self.assertEqual(
```
Should we use the same style of asserts in test? I would opt for `assert x == y`

takunnithan replied:
Yes. Replaced `assertEqual` with `assert`

takunnithan committed:
1. Updates in dag_run_schema  - removed default marshmallow args
2. Removed duplicate code - session related code in test
3. Repladed assertEqual with assert

takunnithan committed:
Merge branch 'master' into issue-9112

ephraimbuddy commented on airflow/api_connexion/schemas/dag_run_schema.py:
```diff
@@ -80,15 +80,15 @@ class Meta:
+    start_date_gte = fields.DateTime(missing=None)
+    start_date_lte = fields.DateTime(missing=None)
+    end_date_gte = fields.DateTime(missing=None)
+    end_date_lte = fields.DateTime(missing=None)
```
Do you have any problem when you remove `missing=None`?

takunnithan replied:
The fields were missing in the deserialized data, when `missing=None` was removed.

takunnithan committed:
Fixed linting issues - removed unused args & other minor fixes

ephraimbuddy reviewed (approved):
LGTM

turbaszek commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -68,40 +70,69 @@ def get_dag_runs(session, dag_id, start_date_gte=None, start_date_lte=None,
+    if data["dag_ids"]:
+        query = query.filter(DagRun.dag_id.in_(data["dag_ids"]))
+
+    dag_runs, total_entries = _fetch_dag_runs(query, session, data["end_date_gte"], data["end_date_lte"],
+                                              data["execution_date_gte"], data["execution_date_lte"],
+                                              data["start_date_gte"], data["start_date_lte"],
+                                              data["page_limit"], data["page_offset"])
```
In `api_connexion/parameters.py` we have some hardcoded params like those. Should we extend the list and use them here instead of using strings?

ephraimbuddy replied:
It no longer exists but can be added back.

turbaszek replied:
Oh indeed, I'm ok with not adding this back

takunnithan committed:
Merge branch 'master' into issue-9112

takunnithan committed:
Fixed the mistake occured in merge conflict resolution

takunnithan committed:
Merge branch 'master' into issue-9112

takunnithan committed:
Fixed Linting issues

takunnithan committed:
Fixed failing test cases

ephraimbuddy commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -87,41 +87,68 @@ def get_dag_runs(
+    try:
+        data = dagruns_batch_form_schema.load(body)
+    except ValidationError as err:
+        raise BadRequest(detail=err.messages)
```
```suggestion
        raise BadRequest(detail=str(err.messages))
```

takunnithan committed:
Converting error message to string type

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>

takunnithan committed:
Test case fix

takunnithan committed:
Trigger CI builds
